metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java

```diff
@@ -92,4 +120,261 @@ private static String getDateFromChoice(Choice choice) {
   private static boolean isEuropeanaProxy(ProxyType proxy) {
     return proxy.getEuropeanaProxy() != null && proxy.getEuropeanaProxy().isEuropeanaProxy();
   }
+
+  /**
+   * Calculates the Europeana Completeness.
+   *
+   * It gives a rank from 0 to maximum 10 for a record. That ranking is divided in two parts.
+   * <p>Up to 5 points for tags and up to 5 points for free-text fields(title, description).
+   * Tags are populated from the first value encountered from each corresponding field.
+   * Free-text fields Title and Description are all taken into account.
+   * </p>
+   * <p>Records with thumbnail or both (title and description) missing get rank 0.<p/>
+   * <p>Points are calculated in the following manner:</p>
+   * <ul>
+   * <li>For tags the number of words inside all fields is calculated and divided by 1. Max value
+   * of points is 5</li>
+   * <li>For free-text the number of words inside all fields is calculated and divided by 5. Max
+   * value of points is 5</li>
+   * </ul>
+   *
+   * @param providerProxy the provider proxy
+   * @param aggregation the provider aggregation
+   * @return the points awarded to the record
+   */
+  private static int computeEuropeanaCompleteness(final ProxyType providerProxy,
+      final Aggregation aggregation) {
+    List<String> tags = new ArrayList<>();
+    List<String> descriptions = new ArrayList<>();
+    List<String> titles = new ArrayList<>();
+    List<Choice> choices = providerProxy.getChoiceList();
+    if (choices != null) {
+      Map<Class<? extends ResourceOrLiteralType>, String> uniqueResourceOrLiteralTypeClassesMap = createCollectionsForResourceOrLiteralType(
+          choices, descriptions);
+      Map<Class<? extends LiteralType>, String> uniqueLiteralTypeClassesMap = createCollectionsForLiteralType(
+          choices, titles);
+      addResourceOrLiteralTypeFromMapsToList(uniqueResourceOrLiteralTypeClassesMap,
+          uniqueLiteralTypeClassesMap, tags);
+    }
+
+    String thumbnailUrl = Optional.ofNullable(aggregation.getObject())
+        .map(ResourceType::getResource)
+        .orElse(null);
+
+    return completenessCalculation(thumbnailUrl, titles, descriptions, tags);
+  }
+
+  private static Map<Class<? extends ResourceOrLiteralType>, String> createCollectionsForResourceOrLiteralType(
```

Comment:
We have the exact same problem in other locations like the ProxyFieldInput.
I don't see another way to generalize this, unless we actually change jibx Choice classes or using reflection which I personally try to avoid...
I created 2 distinct methods one for ResourceOrLiteralType and one for LiteralType to reduce the complexity of having one bigger method.